### ISSUE TITLE
Why do you send the delete to every actor on a server and not just a singl actor?

### ISSUE THREAD

**Issue by wmurphyrd** [mastodon/mastodon#25041]
### Steps to reproduce the problem

@ClearlyClaire reported in immers-space/activitypub-express#94 that Mastodon sends a duplicate Delete activity for reach actor on a remote server if that server doesn't implement shared inbox. 

Given that 1) [shared inbox is optional in ActivityPub ](https://www.w3.org/TR/activitypub/#shared-inbox-delivery) and  2) the [Delete activity side effects in ActivityPub are clearly idempotent](https://www.w3.org/TR/activitypub/#delete-activity-inbox), what's the point of making all those duplicate deliveries?

### Expected behaviour

Do not send duplicate network traffic

### Actual behaviour

sent duplicate network traffic

### Specifications

activitypub-express v4.2.2

**Comment by trwnh** [comment]
there is no guarantee that any two actors on the same server are sharing data. if you send a Delete to one actor but not the other, then the first actor might clear out data but the second actor might not, despite them both being on the same server. "being on the same server" doesn't mean anything unless the server is following the "instance" model.

**Comment by nightpool** [comment]
Correct, the only way to know if actors are sharing an underlying database is if they're using `sharedInbox`. Otherwise, ActivityPub assumes that all actors are not sharing an underlying database. "Remote server" is not a concept in the ActivityPub spec.

**Comment by wmurphyrd** [comment]
> there is no guarantee that any two actors on the same server are sharing data. if you send a Delete to one actor but not the other, then the first actor might clear out data but the second actor might not, despite them both being on the same server. "being on the same server" doesn't mean anything unless the server is following the "instance" model.

@trwnh there is a guarantee right here (emphasis added):

>7.4 Delete Activity

>The side effect of receiving this is that (assuming the object is owned by the sending actor / server) the **server** receiving the delete activity SHOULD remove its representation of the object with the same id, and MAY replace that representation with a Tombstone object.

It says the server must remove the data; not the individual users' data store


>Correct, the only way to know if actors are sharing an underlying database is if they're using sharedInbox. Otherwise, ActivityPub assumes that all actors are independent of all other actors. **"Remote server" is not a concept in the ActivityPub spec.**

@nightpool Please refer to section 7

>7. Server to Server Interactions

>Servers communicate with other servers

**Comment by trwnh** [comment]
you don't know that two actors share a server, even if they share a domain name.

**Comment by erincandescent** [comment]
Sharing a domain name implies sharing an administrative/security domain, but nothing about the database

The only reliable deduplication approach is sharedInbox

**Comment by nightpool** [comment]
> It says the server must remove the data; not the individual users' data store
@wmurphyrd Thanks, that's a good point and I hadn't considered that implication of the spec before. There are a few other references as well:

>  Publicly dereferencable URIs, such as HTTPS URIs, with their authority belonging to that of their **originating server.** (Publicly facing content SHOULD use HTTPS URIs).

(emphasis mine) 

All I can say is that standard explanation and practice is to treat all actors as separate until proven the same, and even the use of sharedInbox for non-public activities is pretty controversial. Generally, when implementers omit `sharedInbox`, it's because they want to be able to process all activities separately for each user. 

Additionally, it's not clear to me how this requirement ("authority belonging to their server") would generalize for non-HTTPS URLs, like DID urls. How do you define the "server" in that case, when every DID url has its own key-based authority? My understanding is that the SocialWG went through a lot of work to try and root out all of these "shared server" assumptions out of the spec, because they weren't compatible with future goals for more decentralized servers. I think this may have just been a section they overlooked.
